ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -265,45 +294,104 @@ void checkRateLimit() {
+    final List<List<?>> tableRows = new ArrayList<>();
+    final List<LogicalSchema> schemas = new ArrayList<>();
+    List<KsqlLocation> remainingLocations = ImmutableList.copyOf(locations);
+    for (int round = 0; ; round++) {
+      final Map<KsqlNode, List<KsqlLocation>> groupedByHost
+          = groupByHost(statement, remainingLocations, round);
+
+      final Map<KsqlNode, Future<PullQueryResult>> futures = new LinkedHashMap<>();
+      for (Map.Entry<KsqlNode, List<KsqlLocation>> entry : groupedByHost.entrySet()) {
+        final KsqlNode node = entry.getKey();
+        final PullQueryContext pullQueryContext = contextFactory.apply(entry.getValue());
+
+        futures.put(node, executorService.submit(() ->  {
+          final TableRows rows = routeQuery.routeQuery(
+              node, statement, executionContext, serviceContext, pullQueryContext);
+          final Optional<List<KsqlNode>> debugNodes = Optional.ofNullable(
+              routingOptions.isDebugRequest()
+                  ? Collections.nCopies(rows.getRows().size(), node) : null);
+          return new PullQueryResult(rows, debugNodes);
+        }));
+      }
+
+      final ImmutableList.Builder<KsqlLocation> nextRoundRemaining = ImmutableList.builder();
+      for (Map.Entry<KsqlNode, Future<PullQueryResult>>  entry : futures.entrySet()) {
+        final Future<PullQueryResult> future = entry.getValue();
+        final KsqlNode node = entry.getKey();
+        try {
+          final PullQueryResult result = future.get();
+          result.getSourceNodes().ifPresent(sourceNodes::addAll);
+          schemas.add(result.getTableRows().getSchema());
+          tableRows.addAll(result.getTableRows().getRows());
+        } catch (ExecutionException e) {
+          LOG.debug("Error routing query {} to host {} at timestamp {} with exception {}",
+              statement.getStatementText(), node, System.currentTimeMillis(), e.getCause());
+          nextRoundRemaining.addAll(groupedByHost.get(node));
+        }
+      }
+      remainingLocations = nextRoundRemaining.build();
+
+      if (remainingLocations.size() == 0) {
+        validateSchemas(schemas);
         return new PullQueryResult(
-            routeQuery(node, statement, executionContext, serviceContext, pullQueryContext),
-            debugNode);
-      } catch (Exception t) {
-        LOG.debug("Error routing query {} to host {} at timestamp {} with exception {}",
-                  statement.getStatementText(), node, System.currentTimeMillis(), t);
+            new TableRows(statement.getStatementText(), queryId, Iterables.getLast(schemas),
+                tableRows),
+            sourceNodes.isEmpty() ? Optional.empty() : Optional.of(sourceNodes));
       }
     }
-    throw new MaterializationException(String.format(
-        "Unable to execute pull query: %s", statement.getStatementText()));
+  }
+
+  private static Map<KsqlNode, List<KsqlLocation>> groupByHost(
```

Comment:
What this method does, is group all locations per the same host, which if round=0, will be the active. So, all locations (all keys) that have the same host as active will we grouped together. Then, in the second round, for any keys that the active failed, we will get the standby that is second in ordering.